Timeout implementation for connections.

Ali-Nasrolahi opened the pull request:
## Objective

Automatically closing a connection, which becomes stalled for some reason.
Fixes MortezaBashsiz/nipovpn#79

## Implementation Details

Before every read/write operation, a timer (`deadline_timer`) gets started/reseted, and after each successful I/O; the timer gets cancelled.  When timer's handlers called ,in case of an expiration, it closes associated socket with the connection; by doing so, any blocking I/O on respected socket will be interrupted.

This patch only addresses *TCPConnection* and *TCPClient* classes' read and write methods. The underlying implementation is identical for both classes.

## Test

For testing, I did following scenarios while `timeout` is specified:

- Grabbing `www.google.com` using `curl` to check to ensure no data transfer issue.
- `telnet` to agent's socket and waiting to be closed by timeout.
- Using a simple python TCP socket as a destination server and never sending any data. Which caused the connection to be timeout-ed.

My tests are not addressing every block of read and write code, therefore this patch should be tested further to ensure stability. 

## About the diffs

Timeout configuration is identical to `timeWait` and uses similar value range.  
I have implemented three main method to handle the timer: `resetTimeout`, `cancelTimeout` and `onTimeout`. 
And few other minor changes.



Please, let me know of any issue/suggestion :)

Ali-Nasrolahi committed:
Timeout implementation for connections.

Objective:
    Automatically closing a connection, which becomes stalled for some reason.

Implementation Details:
    Before every read/write operation, a timer (`deadline_timer`) gets
    started/reseted, and after each successful I/O; the timer gets cancelled.
    When timer's handlers called ,in case of an expiration, it closes associated
    socket with the connection; by doing so, any blocking I/O on respected socket
    will be interrupted.

    This patch only addresses *TCPConnection* and *TCPClient* classes' read and write
    methods. The underlying implementation is identical for both classes.

Fixes MortezaBashsiz/nipovpn#79

MortezaBashsiz commented on core/src/tcpclient.hpp:
```diff
@@ -117,7 +124,7 @@ class TCPClient : public boost::enable_shared_from_this<TCPClient> {
     boost::asio::streambuf writeBuffer_;     // Buffer for data to be written
     boost::asio::streambuf readBuffer_;      // Buffer for data to be read
     boost::asio::ip::tcp::resolver resolver_;// Resolver for endpoint resolution
-    boost::asio::deadline_timer timer_;      // Timer for managing timeouts
```
Good point to remove unused variable

MortezaBashsiz commented on core/src/tcpclient.cpp:
```diff
@@ -197,3 +205,36 @@ void TCPClient::doRead() {
+        return;
+
+    // Start/Reset the timer and cancel old handlers
+    timeout_.expires_from_now(boost::posix_time::milliseconds(config_->general().timeout));
```
Please consider timeout as seconds and not milliseconds

MortezaBashsiz commented on core/src/tcpconnection.cpp:
```diff
@@ -13,6 +13,7 @@ TCPConnection::TCPConnection(boost::asio::io_context &io_context,
       io_context_(io_context),// Store the I/O context
       client_(client),        // Store the TCP client pointer
       timer_(io_context),     // Initialize the deadline timer with the I/O context
```
Please also consider removing useless timer_ from tcpconnection as well

MortezaBashsiz commented on core/src/tcpconnection.hpp:
```diff
@@ -66,14 +66,22 @@ class TCPConnection : public boost::enable_shared_from_this<TCPConnection> {
-            writeBuffer_;              // Buffers for reading and writing data
-    boost::asio::deadline_timer timer_;// Timer used for managing timeouts
+            writeBuffer_;                // Buffers for reading and writing data
+    boost::asio::deadline_timer timer_;  // Timer used for managing timeouts
```
Please also consider removing useless timer_ from tcpconnection as well

MortezaBashsiz commented on nipovpn/etc/nipovpn/config.yaml:
```diff
@@ -1,15 +1,19 @@
+  # Defines the time wait between each repeat to read from socket.
+  #   This directive will be useful when you expect to read very long stream from socket
   timeWait: 0
+  # timeout: unsigned int
```
If you consider timeout as seconds, I think `unsigned int` is too big

MortezaBashsiz commented on core/src/config.hpp:
```diff
@@ -18,6 +18,7 @@ class Config : private Uncopyable {
         std::string fakeUrl;
         std::string method;
         unsigned int timeWait;
+        unsigned int timeout;
```
Please consider timeout as seconds. In this case `unsigned int` is too big

MortezaBashsiz commented on core/src/tcpconnection.cpp:
```diff
@@ -166,3 +177,36 @@ void TCPConnection::doWrite() {
+        timeout_.cancel();
+}
+
+void TCPConnection::onTimeout(const boost::system::error_code &e) {
```
Just for sake of integrity please consider using `&error` instead of `&e`

MortezaBashsiz commented on core/src/tcpconnection.hpp:
```diff
@@ -66,14 +66,22 @@ class TCPConnection : public boost::enable_shared_from_this<TCPConnection> {
+    // Cancel the timeout
+    void cancelTimeout();
+    // Handler in case of a timeout expiration
+    void onTimeout(const boost::system::error_code &e);
```
Same `&error` instead of `&e`

MortezaBashsiz commented on core/src/config.cpp:
```diff
@@ -16,6 +16,7 @@ Config::Config(const RunMode &mode, const std::string &filePath)
       general_({configYaml_["general"]["fakeUrl"].as<std::string>(),
                 configYaml_["general"]["method"].as<std::string>(),
                 configYaml_["general"]["timeWait"].as<unsigned int>(),
+                configYaml_["general"]["timeout"].as<unsigned int>(),
```
Please consider using timeout as seconds, in this case `unsigned int` is too bit

MortezaBashsiz commented on core/src/tcpclient.cpp:
```diff
@@ -20,7 +20,7 @@ TCPClient::TCPClient(boost::asio::io_context &io_context,
       io_context_(io_context),
       socket_(io_context),
       resolver_(io_context),
-      timer_(io_context) {}
```
Good point to remove useless items

MortezaBashsiz commented on core/src/tcpclient.cpp:
```diff
@@ -197,3 +205,36 @@ void TCPClient::doRead() {
+        timeout_.cancel();
+}
+
+void TCPClient::onTimeout(const boost::system::error_code &e) {
```
The same, please consider `&error` instead of `&e`

MortezaBashsiz reviewed (commented):
Thanks a lot for this valuable contribution
There is also a performance degradation with this implementation.
We will come to it when you made these changes

But for your information there are differences with timeout 0 and timeout 10000

```
# When it is 10000
[~]>$ time curl -L https://speed.cloudflare.com/__down?bytes=1 
0
real	0m0.332s
user	0m0.035s
sys	0m0.008s

# When it is 0
[~]>$ time curl -L https://speed.cloudflare.com/__down?bytes=1 
0
real	0m0.165s
user	0m0.041s
sys	0m0.007s
[~]>$ 
```

Ali-Nasrolahi committed:
Minor modifications with a few code cleanup.

Details:
    - Decreasing the timeout value range to 2 bytes.
    - Using seconds instead of miliseconds for timeout indication.
    - Removing unused variable and a couple of variable renaming.

Ali-Nasrolahi commented:
Sure!

I have modified the patch based on your requested changes.

Also, I believe the performance decrement is due to the frequency of `start/reset`-ing and `cancel`-ing the timeout's handler  during each read/write operation. Which as you have mentioned, needs to be improved.
